Inplace forward AD formula optimization

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74205
* #74042
* __->__ #74017

Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

Differential Revision: [D34851563](https://our.internmc.facebook.com/intern/diff/D34851563)

soulitzer committed:
Inplace forward AD formula optimization

[ghstack-poisoned]

soulitzer committed:
Update on "Inplace forward AD formula optimization"




[ghstack-poisoned]

soulitzer committed:
Update on "Inplace forward AD formula optimization"




[ghstack-poisoned]

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -335,9 +335,34 @@ def repl(m: Match[str]) -> str:
+                    #
+                    # (2) may seem too strict, but currently the only ops that satisfy (1) also satisfy (2)
+                    # If there is a need, we can relax (2) to allow any op that has an in-place variant
+                    # We can also avoid hard coding self_t, if there is a need, but that is more work.
```
No need, the inplace modified Tensor is always "self" :)

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -335,9 +335,34 @@ def repl(m: Match[str]) -> str:
+                    # If there is a need, we can relax (2) to allow any op that has an in-place variant
+                    # We can also avoid hard coding self_t, if there is a need, but that is more work.
+                    is_single_method_on_self_t = False
+                    match = re.fullmatch(r'self_t.([\w]*)\((.*)\)', formula)
```
I think you want to add a `^` at the beginning and `$` at the end of the regex to make sure that you don't have something like `self_t.foo(bar) + self_t.bar(baz)` that matches.

soulitzer replied:
Yup, but I used fullmatch instead of match here, so that should already apply automatically?
```
>>> import re
>>> re.fullmatch(r"a", "ab")
>>> re.fullmatch(r"a", "ab") is None
True
>>> re.match(r"a", "ab") is None
False
>>> 
```

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -335,9 +335,34 @@ def repl(m: Match[str]) -> str:
+                        # We want to...
+                        #   Match: self_t.op1(other_p.op2(arg))
+                        #   Avoid: self_ self_t.op1(args) + self_t.op2(args)
+                        # We check this by making sure the first parens in second group is not a closing parens
```
This won't be necessary once you update the regex above

soulitzer replied:
The issue is that I used `.*` in the second group, so something like  `args) + self_t.op2(args` could match. 

One alternative method to what I'm currently doing is using `[^\(\)]*` as the pattern for the second group, but that would prevent us from matching something with nested parenthesis like `self_t.op1(other_p.to(self_p.dtype))`.

albanD replied:
The thing is that the check done here will fail in that case as well if you have `self_t.op1(other_p.to(self_p.dtype)) + self_t.op2(args)`

I feel like the simplest way to do this here is to just manually traverse the string in a for loop and check that parenthesis are balanced.

soulitzer replied:
Good catch, updated!

albanD reviewed (commented):
Sounds good.
Just a small update on the regex to simplify the code.

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

[ghstack-poisoned]

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

[ghstack-poisoned]

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

[ghstack-poisoned]

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -351,9 +351,20 @@ def repl(m: Match[str]) -> str:
+                                        return False
+                                if ch == "(":
+                                    level += 1
+                            assert level == 1, "Parenthesis aren't balanced"
```
nit: avoid plain asserts

soulitzer replied:
Just ended up removing this since this will be caught by the compiler eventually anyway

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851563).

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

Differential Revision: [D34851563](https://our.internmc.facebook.com/intern/diff/D34851563)

[ghstack-poisoned]

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -335,9 +335,44 @@ def repl(m: Match[str]) -> str:
+                        # We want to...
+                        #   Match: self_t.op1(other_p.op2(arg))
+                        #   Avoid: self_t.op1(args) + self_t.op2(args)
+                        #   Avoid: self_t.op1(other_p.op2(arg))+ self_t.op2(args)
```
nit space before the +

albanD reviewed (approved):
Thanks!

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

Differential Revision: [D34851563](https://our.internmc.facebook.com/intern/diff/D34851563)

[ghstack-poisoned]

soulitzer committed:
Update on "Inplace forward AD formula optimization"


Fixes: https://github.com/pytorch/pytorch/issues/71987

Codegen output for `masked_fill__Tensor`:
```
  c10::optional<at::Tensor> self_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto value_t_raw = toNonOptFwGrad(value);
      auto value_tensor = toNonOptTensor(value);
      auto value_t = (value_t_raw.defined() || !value_tensor.defined())
        ? value_t_raw : at::_efficientzerotensor(value_tensor.sizes(), value_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.masked_fill_(mask, value_t) : self_t.masked_fill(mask, value_t);
  }
  if (self_new_fw_grad_opt.has_value() && self_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    self._set_fw_grad(self_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ true);
  }
```

Differential Revision: [D34851563](https://our.internmc.facebook.com/intern/diff/D34851563)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge this
